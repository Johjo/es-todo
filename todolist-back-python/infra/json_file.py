import json
from pathlib import Path
from typing import Any


class JsonFile:
    def __init__(self, path: Path):
        self._path = path

    def insert(self, key: str, value: Any) -> None:
        d = {**self._load_json(), key: value}
        self._path.write_text(json.dumps(d))

    def _load_json(self) -> dict:
        try:
            return json.loads(self._path.read_text())
        except FileNotFoundError:
            return {}

    def read(self, key: str) -> dict:
        values = self._load_json()
        return values[key]

    def all_keys(self) -> list[str]:
        values = self._load_json()
        return list(values.keys())
