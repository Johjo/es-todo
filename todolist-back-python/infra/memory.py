from uuid import UUID

from expression import Option, Nothing, Some

from hexagon.todolist.aggregate import TodolistSnapshot, TaskSnapshot


class Memory:
    def __init__(self) -> None:
        self.all_todolist: dict[str, TodolistSnapshot] = {}

    def by(self, todolist_name: str) -> Option[TodolistSnapshot]:
        if todolist_name not in self.all_todolist:
            return Nothing
        return Some(self.all_todolist[todolist_name])

    def save(self, todolist: TodolistSnapshot):
        self.all_todolist[todolist.name] = todolist

    def task_by(self, todolist_name: str, task_key: UUID) -> TaskSnapshot:
        all_tasks = self.all_todolist[todolist_name].tasks
        task = [task for task in all_tasks if task.key == task_key][0]
        return task

    def all_todolist_by_name(self) -> list[str]:
        return [name for name in self.all_todolist.keys()]

    def all_tasks(self, todolist_name: str) -> list[TaskSnapshot]:
        if todolist_name not in self.all_todolist:
            return []
        return list(self.all_todolist[todolist_name].tasks)


