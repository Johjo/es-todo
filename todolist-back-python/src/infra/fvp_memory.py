from collections import OrderedDict

from src.hexagon.fvp.aggregate import FvpSnapshot
from src.hexagon.shared.type import TaskKey


class FvpMemory:
    def __init__(self) -> None:
        self._snapshot: FvpSnapshot = FvpSnapshot(OrderedDict[TaskKey, TaskKey]())

    def save(self, snapshot: FvpSnapshot):
        self._snapshot = snapshot

    def feed(self, snapshot: FvpSnapshot):
        self._snapshot = snapshot

    def by(self):
        return self._snapshot
