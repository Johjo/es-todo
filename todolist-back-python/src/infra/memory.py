from uuid import UUID

from expression import Option, Nothing, Some

from src.hexagon.todolist.aggregate import TodolistSnapshot, TaskSnapshot


class Memory:
    def __init__(self) -> None:
        self.all_todolist: dict[tuple[str, str], TodolistSnapshot] = {}

    def by(self, user_key: str, todolist_name: str) -> Option[TodolistSnapshot]:
        print(self.all_todolist)
        print(user_key, todolist_name)
        if (user_key, todolist_name) not in self.all_todolist:
            return Nothing
        return Some(self.all_todolist[(user_key, todolist_name)])

    def save(self, user_key: str, todolist: TodolistSnapshot):
        self.all_todolist[(user_key, todolist.name)] = todolist

    def task_by(self, user_key: str, todolist_name: str, task_key: UUID) -> TaskSnapshot:
        all_tasks = self.all_todolist[(user_key, todolist_name)].tasks
        task = [task for task in all_tasks if task.key == task_key][0]
        return task

    def all_todolist_by_name(self, user_key: str) -> list[str]:
        return [name for (todolist_user_key, name) in self.all_todolist.keys() if todolist_user_key == user_key]

    def all_tasks(self, user_key: str, todolist_name: str) -> list[TaskSnapshot]:
        if (user_key, todolist_name) not in self.all_todolist:
            return []
        return list(self.all_todolist[(user_key, todolist_name)].tasks)

    def __repr__(self):
        return str(self.all_todolist)


