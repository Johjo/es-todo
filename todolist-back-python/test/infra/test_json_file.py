from pathlib import Path

import pytest
from approvaltests import verify  # type: ignore
from approvaltests.reporters import PythonNativeReporter

from infra.json_file import JsonFile

@pytest.fixture
def json_path(tmp_path) -> Path:
    return tmp_path / "test_todolist.json"

def test_insert_one_value(json_path: Path):
    something = JsonFile(json_path)
    something.insert("my_todolist", {"name": "my_todolist", "key": "value"})

    verify(json_path.read_text(), reporter=PythonNativeReporter())


def test_insert_many_values(json_path: Path):
    something = JsonFile(json_path)
    something.insert("todolist 1", {"name": "todolist 1", "key": "value"})
    something.insert("todolist 2", {"name": "todolist 2", "key": "value"})

    verify(json_path.read_text(), reporter=PythonNativeReporter())

def test_read_value(json_path: Path):
    something = JsonFile(json_path)
    something.insert("todolist 1", {"name": "todolist 1", "key": "value"})
    something.insert("todolist 2", {"name": "todolist 2", "key": "value"})

    assert something.read("todolist 2") == {"name": "todolist 2", "key": "value"}
