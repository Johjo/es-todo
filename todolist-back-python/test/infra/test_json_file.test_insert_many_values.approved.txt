{"todolist 1": {"name": "todolist 1", "key": "value"}, "todolist 2": {"name": "todolist 2", "key": "value"}}
